File: 2025/day07-1.hs
import Data.List (elemIndex)
import Data.Maybe (fromJust)
import Data.Set (fromList, toList)

solution :: String -> Int
solution content = snd . foldl propagate ([findStart . head . lines $ content], 0) $ (filter (any (/= '.')) . tail . lines $ content)

propagate :: ([Int], Int) -> String -> ([Int], Int)
propagate (beams, count) line =
  ( split beams splitters,
    count + (length . filter (`elem` splitters) $ beams)
  )
  where
    splitters = findSplitters line

split :: [Int] -> [Int] -> [Int]
split beams splitters = unique (filter (`notElem` splitters) beams ++ concatMap (\s -> [s - 1, s + 1]) splitters)

findSplitters :: String -> [Int]
findSplitters = map fst . filter ((== '^') . snd) . zip [0 ..]

findStart :: String -> Int
findStart = fromJust . elemIndex 'S'

unique :: [Int] -> [Int]
unique = toList . fromList

main :: IO ()
main = do readFile "day07.dat" >>= print . solution

File: 2025/day07_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    return propagate(
        find_start(content[0]),
        [line for line in content[1:] if any(ch != "." for ch in line)],
    )


def propagate(start: int, content: list[str]) -> int:
    count = 0
    beams = set([start])
    for line in content:
        splitters = find_splitters(line)
        count += len([beam for beam in beams if beam in splitters])
        beams = split(beams, splitters)
    return count


def find_splitters(line: str) -> list[int]:
    return [index for index, char in enumerate(line) if char == "^"]


def split(beams: set[int], splitters: list[int]) -> set[int]:
    return set(
        [beam for beam in beams if beam not in splitters]
        + [splitter - 1 for splitter in splitters]
        + [splitter + 1 for splitter in splitters]
    )


def find_start(line: str) -> int:
    return line.index("S")


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
